### DISCUSSION TITLE
I need json.h and json.cpp to optimize size of builded file

### DISCUSSION THREAD

**Discussion by anhtu812** [nlohmann/json#2965] [Q&A]
I need optimize size of webassembly file that builded by Emscripten (em++) from "main.cpp" (have include json.hpp). I tested and sure that if main.cpp only include json.h and link with json.a (static library build from json.cpp use em++) then webassembly output file reduce size. Optimize size of file of web mobile app is importent.

In the future, can you add that feature? Thank you very much.

**Comment by nlohmann** [comment]
The library heavily depends on templates. I do not think a static library is possible in this case.

**Reply by nlohmann** [reply]
Do you need further assistance with this issue?

**Reply by HJWAJ** [reply]
I’m replacing jsoncpp by nlohmann/json, and have run into same problem. I noticed that each of my cpp's including json.hpp expands, from 10+k to 100+k, which resulting my app size grow ~2M. Most of my code should be using same templates, so I suspect there may be much repeating code in each of the compilation units, which could be avoided by seperating definition and implementation.

I need help, how can I convince or overthrow my suspecion? What can I do to shrink my lib size? Thank you very much!

**Reply by nlohmann** [reply]
There is json_fwd.hpp for forward declarations.

**Reply by HJWAJ** [reply]
Yes, but it's just for forward declarations. I think it is used for

#include "json_fwd.hpp"
// do something custom
#include "json.hpp"

But what I want is something maybe like a "json_without_implementation.hpp" and a "json_impl.cpp".

Am I missing something?

**Reply by falbrechtskirchinger** [reply]
Sure, each compilation unit may instantiate the same templates but at link-time only one of the instantiations will make it into the output (library/executable). Depending on your compiler and/or platform you have different options to reduce binary size. For example, have you tried LTO?

**Reply by HJWAJ** [reply]
I thought at link time only one instantiations would be reserved, but the fact seems to be different. I also tried -flto and got another 3MB lib size grow :-(
BTW, I'm using apple clang 13 on MacOS and iOS with Xcode.

**Reply by nlohmann** [reply]
Strange. Can you share the project?

**Reply by gregmarr** [reply]
> Yes, but it's just for forward declarations. I think it is used for

    #include "json_fwd.hpp"
    // do something custom
    #include "json.hpp"

Not necessarily.  It is generally used whenever you need to have a pointer or reference to an object to pass around, but aren't actually going to be using the data or functions of the object.  Since this is primarily template code, it really isn't feasible to separate the implementation into a separate file, as nlohmann said earlier.

Are you perhaps building a static lib rather than an executable?

**Reply by HJWAJ** [reply]
Sorry for the late reply. The project is too complex to share, and I failed to make a minimal reproducible demo. 

I‘m building a static lib. After digging into the disassembled code from the lib(with objdump), I find different .cpp.o disassembled contains different part of json code. Almost full code is packed into the lib. However, when building an executable, it seems code only used by the executable is packed.

So yes, now I think split `json.hpp` into `json.h` and `json.cpp` will not help. I deleted some unused code by hand, and get a lib size shrink.

However, I still don't know why `-flto` makes my lib larger.

**Reply by gregmarr** [reply]
In general, building a static library does not do optimization across .o files.  It just assembles the individual .o files into a .a archive.  The duplicate code will be eliminated when the library is linked into the final binary.  This is outside the scope of this library, but there are a variety of questions about using LTO with static libraries on Stack Overflow, with some potential solutions that you may find helpful.

**Reply by weiqingtangx** [reply]
I think json.h and json.cpp still will help. In my Case, I own a static library as  a bussiness line, each bussiness line have size limitations of static library, sometimes my static library cant not into the integration due to the limitation. So the smaller static library the better.

**Reply by falbrechtskirchinger** [reply]
@weiqingtangx I dare you (or anyone) to develop an example that showcases the effects you suggest a split into header and source file(s) would have (excluding special compiler and linker flags). Remember we're talking about heavily templated code and that the linker (by default) doesn't remove unused symbols from static libraries.

Until someone takes me up on this challenge and proves otherwise, I'll remain skeptical of any significant effects. Good luck! :-)

**Reply by deepbluev7** [reply]
While a static library alone probably is not very useful, using extern template could improve the compilation time of the default template argument by a lot. This however also requires a split into multiple files, you can't define your template member functions inline anymore and it could actually be built as a static library.

**Reply by falbrechtskirchinger** [reply]
This is off-topic since OP asked about binary size and not compilation time, but I'll engage anyway:

What is the "default template," practically speaking?

The most basic use of the library requires calling template member functions with a number of plausible types. Explicitly instantiating all of them would be a chore.
Also, I think this could be done now without any changes to the library. All you need is a wrapper header for the explicit instantiation declarations and a source file for the explicit instantiation definitions, right? But if you forget just a single instantiation all benefit vanishes.

**Reply by deepbluev7** [reply]
Well, the default template is just the one you get with the default nlohmann::json object. This instantiates the parser once per translation unit, which then makes the compilation process slow and bloats up the object files and static libraries (but the final link process removes them again, so yes, it has not much effect on the final executable).

If you look at it, `nlohmann::json::parse()` takes mostly the same template arguments each time, but takes up a considerable amount of compilation time (and that is just the instantiation of that template, in total nlohmann json makes up about 50% of the compilation time of my projects):

![grafik](https://user-images.githubusercontent.com/16312324/188849901-bc1065d4-e85b-4782-a83c-e6b9f8d00179.png)

Yes, the .get() functions and such take different types each time, but my library mostly uses the default nlohmann::json parameter and the nlohmann::json::parse function with char* or std::string{,_view} parameters, which each time instantiates a very, very large template machinery.

So what this lib could do is mark those functions as extern template, put the definition and explicit instantiation in a different file and let users separately compile those. That way instead of instantiating the json parser in 200 translation units, it would happen in one, giving my library a 50%*199/200 compilation time speedup.

The problem why this can't be done right now with a wrapper, is that this doesn't work with inline definitions of template functions, since those are instantiated before the extern template declaration. And the template gets instantiated in every translation unit because of the _json UDL, even if you don't use any nlohmann function (although that can somewhat be mitigated with the _fwd header).

And no, as long as you have marked an instantiation as extern template before the first use of that template, you do have the benefit in that translation unit. How exactly to do that in a header only lib might be a bit difficult, but some macro that can be defined to skip template instantiation in the header might work.

**Reply by falbrechtskirchinger** [reply]
> If you look at it, `nlohmann::json::parse()` takes mostly the same template arguments each time, but takes up a considerable amount of compilation time

That is a fair point. There might be other candidates as well (`dump()`?).

> The problem why this can't be done right now with a wrapper, is that this doesn't work with inline definitions of template functions, since those are instantiated before the extern template declaration.

Right. 


I don't think anyone would try to stop anybody from trying to improve compilation times. CI is quite slow these days because of it as well.
How to do it is a different question. The GNU standard library does something related with its template definition files (`*.tcc`) and `_GLIBCXX_EXTERN_TEMPLATE` which might be worth looking into...
